### PR TITLE
information_schema case-insensitivity

### PR DESCRIPTION
**Pull request by acharis**
@sougou @giaquinti 
https://github.com/youtube/vitess/issues/3069
This is what was talked about in #vitess but ... it doesn't work.
If you grep around for information_schema and find some tests, and flip the case, you'll see.
This is WIP both because it needs tests, and because it's not really working.

Secondarily, please also furnish a list of what you consider "system schemas"

### CODE REVIEW

File: go/vt/vtgate/planbuilder/route.go [hunk 1]
```diff
@@ -533,6 +536,11 @@ func (rb *route) Wireup(bldr builder, jt *jointab) error {
 	return nil
 }
 
+func systemTable(qualifier string) bool {
+	lowered := strings.ToLower(qualifier)
+	return lowered == "information_schema" || lowered == "performance_schema" || lowered == "sys"
```
Comment (bbeaudreault) [inline go/vt/vtgate/planbuilder/route.go]: I could be wrong, but I think strings.EqualFold may be more performant here. It avoids creating a new string like ToLower, and can abort early in the string if it's not a match.

Comment (acharis) [reply go/vt/vtgate/planbuilder/route.go]: oooh! i like it

**Comment by acharis** [comment]
Just to comment here instead of only in commit messages, I think this is working now and ready for review/merge.

**Comment by acharis** [comment]
in DM, Guido is bringing up "_vt" and "mysql" schemas as well. thoughts? @sougou

**Comment by giaquinti** [comment]
What do we think about adding as well `mysql`?

File: go/vt/vtgate/planbuilder/route.go [hunk 2]
```diff
@@ -519,10 +520,13 @@ func (rb *route) Wireup(bldr builder, jt *jointab) error {
 				return
 			}
 		case sqlparser.TableName:
-			if node.Qualifier != infoSchema {
+			if !systemTable(node.Qualifier.String()) {
 				node.Name.Format(buf)
 				return
 			}
+			// We accept INFORMATION_SCHEMA, but underlying mysql might not
```
Comment (bbeaudreault) [inline go/vt/vtgate/planbuilder/route.go]: i personally think we should let the underlying mysql do throw an error.  the idea here is vitess should be invisible. but if it doesnt work underlying then it doesn't work. 

Separately, I also think the less we rewrite queries the better, but that may be a lost cause. It's a gotcha when grepping logs and such for patterns you expect.

**Comment by bbeaudreault** [comment]
@giaquinti what's your use-case here? I thought the goal here was to make vitess "Do No Harm" by allowing queries that work in mysql to work in vitess.  This made sense for `INFORMATION_SCHEMA`:

```
mysql> select * from INFORMATION_SCHEMA.INNODB_SYS_FOREIGN limit 1;
Empty set (0.00 sec)

mysql> select * from information_schema.INNODB_SYS_FOREIGN limit 1;
Empty set (0.00 sec)
```

See it works for both upper and lowercase.

I tried the same for performance_schema, mysql, and sys. None of those have the same semantics in MySQL.

If this is about https://dev.mysql.com/doc/refman/5.7/en/server-system-variables.html#sysvar_lower_case_table_names and allowing any table identifier to be case-insenstive then that is a different project with different implications.

File: go/vt/vtgate/planbuilder/route.go (hunk 2 above)
Comment (acharis) [reply go/vt/vtgate/planbuilder/route.go]: i'll let @giaquinti weigh in

**Comment by giaquinti** [comment]
My proposal is to support case insensitive comparisons for system schemas only. 

I think this will help to improve Vitess adoption as there are tons of applications that we don't control that are relying on this feature.

I agree that `_vt` can't be considered a system schema but I'm 👍 to include:
- `information_schema`
- `performance_schema`
- `mysql`
- `sys`  

Regarding the example that you posted I'm not sure I'm following but both of your queries are working fine for me

```
mysql> select count(*) from PERFORMANCE_SCHEMA.status_by_user;
+----------+
| count(*) |
+----------+
|       84 |
+----------+
1 row in set (0.01 sec)

mysql> select count(*) from performance_schema.status_by_user;
+----------+
| count(*) |
+----------+
|       84 |
+----------+
1 row in set (0.00 sec)
```

I'm running the latest Percona 5.7 build via Docker
mysql> SELECT @@lower_case_table_names;
+--------------------------+
| @@lower_case_table_names |
+--------------------------+
|                        1 |
+--------------------------+
1 row in set (0.00 sec)

File: go/vt/vtgate/planbuilder/route.go (hunk 2 above)
Comment (giaquinti) [reply go/vt/vtgate/planbuilder/route.go]: I'm +1 to don't rewrite the query

**Comment by bbeaudreault** [comment]
@giaquinti apologies, I thought we were running default but apparently we are not. We use value of `1` for that, which is why those queries don't work for me. 

Do you have real examples of apps that refer to the other system tables (not information_schemas) by their name capitalized?

Anyway, if we remove the rewriting of the query we can probably agree to ship this.  Case-sensitivity aside, we should still be able to route the `mysql` schema.

**Comment by giaquinti** [comment]
I don't have any examples of apps that refer to other system schemas by their capitalized name off the top of my head. My suggestion is to extend this functionality as well to the other schemas and keep compatibility as close as possible to MySQL. E.g.: if new project uses capitalized name for `performance_schema`, then it will work with a standard MySQL installation and Vitess.

**Comment by sougou** [comment]
AFAIK, vitess tries to preserve the original case of column as well as table names. So, they should just pass through. If there are cases where it doesn't, we should fix those cases.

So, the only issue should be related to special-casing of the system schema names to be case-insensitive, which is fair. And I think it should be ok to include `mysql` in that list, although I have no idea what's in that schema :).

**Comment by sougou** [comment]
So, I think the consensus is to add `mysql` to the list of system schemas? Once that's done, I can submit this.

**Comment by bbeaudreault** [comment]
Agreed, let's remove the `ToLowerCase()` and add `mysql` to the list.

**Comment by sougou** [comment]
Waiting on the final changes for this one.

**Comment by acharis** [comment]
got back around to this one:
added mysql to list of system schemas
removed the ToLower
rebased on master

**Comment by sougou** [comment]
LGTM

**Pull request merged**